### PR TITLE
add javascript filtering

### PR DESCRIPTION
**Pull request by osheroff**
This introduces javascript per-row hooks that are called before passing
a row off to the producer.  This lets users munge data, filter out
output based on arbitrary code, and choose their own kafka topics.  More
possibilities are obviously available, but it seems like a good place to
start.

TODO:  some testing, a bit more documentation.

### CODE REVIEW

File: src/main/java/com/zendesk/maxwell/scripting/WrappedRowMap.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	public String getTable() {
+		return row.getTable();
+	}
+
+	public String getDatabase() {
+		return row.getDatabase();
+	}
+
+	public String getPosition() {
+		BinlogPosition p = row.getPosition().getBinlogPosition();
```
Comment (ashashwat) [inline src/main/java/com/zendesk/maxwell/scripting/WrappedRowMap.java]: NPE if row.getPosition() will ever be NULL.  Not sure if that can ever happen though.

File: docs/docs/config.md [hunk 2]
```diff
@@ -95,6 +95,26 @@ you can use `exclude_columns` to filter out columns by name.
...
+		var username = row.data.get("username");
+		if ( username == "osheroff" )
+			row.suppress();
+
+		row.data.put("username", username.toUpperCase());
+	}
+}
+```
+
+There's a longer example [here]().
```
Comment (ashashwat) [inline docs/docs/config.md]: This link points to nowhere.

File: src/main/java/com/zendesk/maxwell/MaxwellConfig.java [hunk 3]
```diff
@@ -533,6 +539,15 @@ private void setup(OptionSet options, Properties properties) {
 				}
 			}
 		}
+
+		if ( this.javascriptFile != null ) {
```
[the remark this answers is not in the document]
Comment (osheroff) [reply src/main/java/com/zendesk/maxwell/MaxwellConfig.java]: nah, is cool.  let 'em name it whatever.

File: docs/docs/config.md (hunk 2 above)
Comment (osheroff) [reply docs/docs/config.md]: yeah i will fix this up once I get it merged (& get a reliable github link)

**Comment by osheroff** [comment]
@timbertson @liulikun any comments?

**Comment by osheroff** [comment]
pingo pongo

**Comment by liulikun** [comment]
A very interesting approach! I like the idea. Will the performance be an issue?

**Comment by osheroff** [comment]
@liulikun sorry for the delay, I was in Montreal and dropped my laptop out of bed.  I'm going to write a little performance harness for maxwell as it'd be great to get decent benchmarks on stuff like this (as well as other performance work).

**Comment by osheroff** [comment]
ok @liulikun I dusted off this code and started perf-testing it.  The overhead is quite low, something on the order of 2-4% for just entering the interpreter, calling a function, exiting.  Of course more complicated javascript would have different overheads.

**Comment by osheroff** [comment]
@liulikun or @ericwush this is ready with tests and I'd love a look-see.  I wasn't exactly sure what to do with errors from the script engine, so I'm just letting maxwell crash for sure.

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java [hunk 4]
```diff
@@ -151,6 +155,9 @@ public void work() throws Exception {
 		rowCounter.inc();
 		rowMeter.mark();
 
+		if ( scripting != null )
+			scripting.invoke(row);
+
```
Comment (liulikun) [inline src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: Should we skip `processRow` if it's suppressed?

**Comment by liulikun** [comment]
Looks good! We can run some performance tests once it's merged/released.

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java (hunk 4 above)
Comment (osheroff) [reply src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: the thing I'd be worried about is the user shooting themselves in the foot, suppressing everything including maxwell updates, etc.  we have lots of code in the filters to make sure that doesn't happen, but nothing here.

Comment (liulikun) [reply src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: 👍

**Comment by osheroff** [comment]
@liulikun can you do another pass when you get a sec?  I made some changes to allow ddl rows and heartbeat rows work better.

**Comment by liulikun** [comment]
Nice one! 👍

**Pull request merged**